armory/utils/metrics.py

```diff
@@ -431,6 +431,192 @@ def _check_object_detection_input(y, y_pred):
+    """
+
+    IOU_THRESHOLD = 0.5
+    # Precision will be computed at recall points of 0, 0.1, 0.2, ..., 1
+    RECALL_POINTS = np.linspace(0, 1, 11)
+
+    # Converting all boxes to a list of dicts (a list for predicted boxes, and a
+    # separate list for ground truth boxes), where each dict corresponds to a box and
+    # has the following keys "img_idx", "label", "box", as well as "score" for predicted boxes
+    pred_boxes_list = []
+    gt_boxes_list = []
+    for img_idx, (y, y_pred) in enumerate(zip(list_of_ys, list_of_y_preds)):
+        for gt_box_idx in range(len(y["labels"][0].flatten())):
+            label = y["labels"][0][gt_box_idx]
+            box = y["boxes"][0][gt_box_idx]
+
+            gt_box_dict = {"img_idx": img_idx, "label": label, "box": box}
+            gt_boxes_list.append(gt_box_dict)
+
+        for pred_box_idx in range(len(y_pred["labels"].flatten())):
+            label = y_pred["labels"][pred_box_idx]
+            box = y_pred["boxes"][pred_box_idx]
+            score = y_pred["scores"][pred_box_idx]
+
+            pred_box_dict = {
+                "img_idx": img_idx,
+                "label": label,
+                "box": box,
+                "score": score,
+            }
+            pred_boxes_list.append(pred_box_dict)
+
+    # Union of (1) the set of all true classes and (2) the set of all predicted classes
+    set_of_class_ids = set([i["label"] for i in gt_boxes_list]) | set(
+        [i["label"] for i in pred_boxes_list]
+    )
+
+    # Remove the class ID that corresponds to a physical adversarial patch in APRICOT
+    # dataset, if present
+    # TODO: retrieve this ID from datasets.py rather than hardcoding it
+    set_of_class_ids.discard(-10)
+
+    # Initialize dict that will store AP for each class
+    average_precisions_by_class = {}
+
+    # Compute AP for each class
+    for class_id in set_of_class_ids:
+
+        # Buiild lists that contain all the predicted/ground-truth boxes with a
+        # label of class_id
+        class_predicted_boxes = []
+        class_gt_boxes = []
+        for pred_box in pred_boxes_list:
+            if pred_box["label"] == class_id:
+                class_predicted_boxes.append(pred_box)
+        for gt_box in gt_boxes_list:
+            if gt_box["label"] == class_id:
+                class_gt_boxes.append(gt_box)
+
+        # Determine how many gt boxes (of class_id) there are in each image
+        num_gt_boxes_per_img = Counter([gt["img_idx"] for gt in class_gt_boxes])
+
+        # Initialize array where we'll keep track of whether a gt box has been matched to a
+        # prediction yet. This is necessary because if multiple predicted boxes of class_id
+        # overlap with a single gt box, only one of the predicted boxes can be considered a
+        # true positive
+        for img_idx, num_gt_boxes in num_gt_boxes_per_img.items():
+            num_gt_boxes_per_img[img_idx] = np.zeros(num_gt_boxes)
```

Comment:
num_gt_boxes_per_img is type Counter, but what it stores is an array rather than a count.  It works, but I wonder if a regular dict is more appropriate.

Reply:
It initially stores a count, but I agree that once the keys are mapped to np arrays the variable name no longer makes sense. Updating